src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,158 @@
+        } else if (index > 0 && index < size) {
+            Node<T> current = findByIndex(index);
+            current.prev.next = newNode;
+            newNode.next = current;
+            newNode.prev = current.prev;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return findByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = findByIndex(index);
+        T changedValue = currentNode.item;
+        currentNode.item = value;
+        return changedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = findByIndex(index);
+        if (currentNode.prev != null && currentNode.next != null) {
+            currentNode.prev.next = currentNode.next;
+            currentNode.next.prev = currentNode.prev;
+        } else if (currentNode == head && currentNode.next != null) {
+            currentNode.next.prev = null;
+            head = currentNode.next;
+        } else if (currentNode == tail && currentNode.prev != null) {
+            currentNode.prev.next = null;
+            tail = currentNode.prev;
+        }
+        size--;
+        return currentNode.item;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            Node<T> nodeToRemove = findByIndex(i);
+            if ((object == null && nodeToRemove.item == null)
+                    || (object != null && object.equals(nodeToRemove.item))) {
+                if (nodeToRemove.prev != null && nodeToRemove.next != null) {
+                    nodeToRemove.prev.next = nodeToRemove.next;
+                    nodeToRemove.next.prev = nodeToRemove.prev;
+                } else if (nodeToRemove == head && nodeToRemove.next != null) {
+                    nodeToRemove.next.prev = null;
+                    head = nodeToRemove.next;
+                } else if (nodeToRemove == tail && nodeToRemove.prev != null) {
+                    nodeToRemove.prev.next = null;
+                    tail = nodeToRemove.prev;
+                }
+                size--;
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size <= 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index + " out of bounds");
+        }
+    }
+
+    private void checkIndexForAdd(int index) {
```

Comment:
Private methods and classes should be after public ones in your class. Move the `checkIndex` and `checkIndexForAdd` methods below the `Node` inner class.